apis/core/v1alpha1/resource.go

```diff
@@ -171,13 +171,43 @@ type ProviderSpec struct {
 	CredentialsSecretRef *SecretKeySelector `json:"credentialsSecretRef,omitempty"`
 }
 
-// A ProviderConfigSpec defines the common way to get to the necessary objects
-// to connect to the provider.
+// A ProviderConfigSpec defines the desired state of a provider config. A
+// provider config may embed this type in its spec in order to support standard
+// fields. Provider configs may choose to avoid embedding this type as
+// appropriate, but are encouraged to follow its conventions.
 type ProviderConfigSpec struct {
-	// CredentialsSecretRef references a specific secret's key that contains
-	// the credentials that are used to connect to the provider.
+	// Credentials required to authenticate to this provider.
+	Credentials ProviderCredentials `json:"credentials"`
+}
+
+// A CredentialsSource is a source from which provider credentials may be
+// acquired.
+type CredentialsSource string
```

Comment:
I like this pattern! One thing that came to mind as a question is I think that we have discussed before some sort of credentials chain (i.e. if first option doesn't work, use the second, etc.). Do you think that would likely be implemented more at the managed resource level as opposed to the `ProviderConfig` level (i.e. managed resources have a chain of provider config refs rather than `ProviderConfig` having a chain of credentials sources)?

Reply:
Hrm, that's a good question. I feel that an array here could be a good UX for that, but it would increase the blast radius of this change (i.e. the complexity of all connect methods would be increased):

```yaml
apiVersion: aws.crossplane.io/v1beta1
kind: ProviderConfig
metadata:
  name: default
spec:
  # Fetch credentials natively (i.e. via IRSA)
  credentials:
  - source: Native
  # Fall back to fetching credentials from a secret.
  - source: Secret
     secretRef:
       namespace: default
       name: gcp
       key: credentials.json
```

Reply:
Ah - there is a way to implement this in future without a breaking API change. @kasey suggested a distinct `Chained` source. I imagine this could take an array of other sources, e.g.:

```yaml
apiVersion: example.org/v1alpha1
kind: ProviderConfig
metadata:
  name: example
spec:
  credentials:
    source: Chained
    chain:
    - source: Native
    - source: ...
```